src/Microsoft.AspNetCore.Mvc.TagHelpers/Cache/HybridHtmlFragmentCache.cs

```diff
@@ -0,0 +1,219 @@
+                var stringBuilder = new StringBuilder();
+                using (var writer = new StringWriter(stringBuilder))
+                {
+                    content.WriteTo(writer, context.HtmlEncoder);
+                }
+
+                var result = new StringBuilderHtmlContent(stringBuilder);
+
+                return _memoryCache.Set(key, result, options) as IHtmlContent;
+            }
+        }
+
+        public async Task<IHtmlContent> SetDistributedAsync(string key, Func<Task<IHtmlContent>> renderContent, HtmlFragmentCacheContext context, IDistributedCache cache)
+        {
+            var options = new DistributedCacheEntryOptions();
+
+            if (context.ExpiresOn != null)
+            {
+                options.SetAbsoluteExpiration(context.ExpiresOn.Value);
+            }
+
+            if (context.ExpiresAfter != null)
+            {
+                options.SetAbsoluteExpiration(context.ExpiresAfter.Value);
+            }
+
+            if (context.ExpiresSliding != null)
+            {
+                options.SetSlidingExpiration(context.ExpiresSliding.Value);
+            }
+
+            var content = await renderContent();
+
+            var stringBuilder = new StringBuilder();
+            using (var writer = new StringWriter(stringBuilder))
+            {
+                content.WriteTo(writer, context.HtmlEncoder);
+            }
+
+            var serialized = Encoding.UTF8.GetBytes(stringBuilder.ToString());
+
+            await cache.SetAsync(key, serialized, options);
+
+            return content;
+        }
+
+        public Task<IHtmlContent> GetValueAsync(string key, HtmlFragmentCacheContext context)
+        {
+            IDistributedCache distributedCache;
+
+            // is the distributed option set?
+            if (EnsureDistributedService(context, out distributedCache))
+            {
+                return GetDistributedValueAsync(key, distributedCache);
+            }
+            else
+            {
+                IHtmlContent value;
+                if(GetMemoryValue(key, out value))
+                {
+                    return Task.FromResult(value);
+                }
+                else
+                {
+                    return Task.FromResult<IHtmlContent>(null);
+                }
+            }
+        }
+
+        public bool GetMemoryValue(string key, out IHtmlContent value)
+        {
+            return _memoryCache.TryGetValue(key, out value);
+        }
+
+        public async Task<IHtmlContent> GetDistributedValueAsync(string key, IDistributedCache cache)
+        {
+            var encoded = await cache.GetAsync(key);
+
+            if(encoded == null)
+            {
+                return null;
+            }
+
+            var content = Encoding.UTF8.GetString(encoded);
+            return new HtmlEncodedString(content);
+        }
+
+        private bool EnsureDistributedService(HtmlFragmentCacheContext context, out IDistributedCache distributedCache)
```

Comment:
I named it like this before calling it this way, because it can throw an exception, and AFAIK the `TryXYZ` pattern is not supposed to throw anything.